LLMNR client and responder support

jukkar opened the pull request:
This adds link-layer multicast name resolution (LLMNR) support. LLMNR is used in Windows to resolve hostnames. See RFC 4795 for details. The PR adds support for both client (zephyr makes LLMNR query) and responder (windows makes a LLMNR query and zephyr responds to it).

This works well with #561 and net-usb support. Tested with Windows 8.1.

jukkar committed:
net: llmnr: Add link-local mcast name resolution client support

This implements LLMNR client from RFC 4795. This means that caller
is able to resolve DNS resource records using multicast DNS.
The LLMNR is used in Windows networks.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: dns: Add LLMNR client support to DNS resolver

By default the LLMNR is disabled in this sample. You can enable
it by setting CONFIG_LLMNR_RESOLVER=y

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: llmnr: Add LLMNR responder support

This allows zephyr to listen LLMNR DNS queries sent by Windows
and respond to them. See RFC 4795 for details.

The feature requires that hostname is set properly to the
zephyr device and LLMNR is configured properly.

Typically following config options are enough for this support:

CONFIG_NET_HOSTNAME_ENABLE=y
CONFIG_NET_HOSTNAME="zephyr-device"
CONFIG_DNS_RESOLVER=y
CONFIG_LLMNR_RESPONDER=y

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

dbkinder commented on samples/net/dns_resolve/README.rst:
```diff
@@ -17,6 +17,10 @@ The multicast DNS (mDNS) client resolver support can be enabled by setting
 See https://tools.ietf.org/html/rfc6762 for more details about mDNS.
 
+The link-local multicast name resolution (LLMNR) client resolver support can be
+enabled by setting CONFIG_LLMNR_RESOLVER Kconfig variable.
```
change to this (to get a cross reference to the config option documentation):

    enabled by setting the :option:`CONFIG_LLMNR_RESOLVER` Kconfig option.

jukkar committed:
samples: net: dns: Cross reference mDNS config option

Set CONFIG_MDNS_RESOLVER to get a cross reference to the config
option documentation

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

dbkinder reviewed (approved):
+1 for docs

jukkar commented:
@pfalcon & @tbursztyka please review whenever you have time. I have tested this with win 8.1 and the PR worked fine so we could apply it.

tbursztyka commented on include/net/dns_resolve.h:
```diff
@@ -62,6 +62,19 @@ enum dns_query_type {
+#else
+#define LLMNR_SERVER_COUNT 0
+#endif /* CONFIG_MDNS_RESOLVER */
+
```
Add a #define DNS_MAX_SERVERS (or something semantically better if you have any idea) which will add the 3 of the server count. Will be easier in relevant code then, no need to make the addition manytimes.

pfalcon replied:
@tbursztyka , do you mean instead of:

~~~
#define DNS_MAX_MCAST_SERVERS (MDNS_SERVER_COUNT + LLMNR_SERVER_COUNT)
} servers[CONFIG_DNS_RESOLVER_MAX_SERVERS + DNS_MAX_MCAST_SERVERS];
~~~

to have:

~~~
#define DNS_MAX_SERVERS (CONFIG_DNS_RESOLVER_MAX_SERVERS + MDNS_SERVER_COUNT + LLMNR_SERVER_COUNT)
} servers[DNS_MAX_SERVERS];
~~~

? (The comment looks misplaced a bit.)

I'm actually not sure which way is better. I'd say, I'd prefer that whenever possible, CONFIG_* options were used directly in various expression, instead of hidden by extra layer of indirection (easier to grep code this way).

tbursztyka replied:
The idea, as you saw in the code, is to avoid repeating the lengthy addition of all these CONFIG_ yes.

tbursztyka commented on subsys/net/lib/dns/Kconfig:
```diff
@@ -20,6 +20,17 @@ config MDNS_RESOLVER
+	bool "Enable LLMNR support"
+	default n
+	help
+	This option enables link local multicast name resolution client side
```
help message should be indented 2 spaces more, to follow all the other ones in this Kconfig

tbursztyka commented on subsys/net/lib/dns/resolve.c:
```diff
@@ -77,15 +82,73 @@ NET_BUF_POOL_DEFINE(dns_qname_pool, DNS_RESOLVER_BUF_CTR, DNS_MAX_NAME_LEN,
+		if (net_is_ipv4_addr_mcast(&net_sin(addr)->sin_addr)) {
+			if (net_sin(addr)->sin_addr.s4_addr[3] == 251) {
+				return true;
+			} else {
```
no need of that else {} here. Actually you could merge the 2 if in one with a && for returning true.

tbursztyka commented on subsys/net/lib/dns/resolve.c:
```diff
@@ -77,15 +82,73 @@ NET_BUF_POOL_DEFINE(dns_qname_pool, DNS_RESOLVER_BUF_CTR, DNS_MAX_NAME_LEN,
+		if (net_is_ipv6_addr_mcast(&net_sin6(addr)->sin6_addr)) {
+			if (net_sin6(addr)->sin6_addr.s6_addr[15] == 0xfb) {
+				return true;
+			} else {
```
same here

tbursztyka commented on subsys/net/lib/dns/resolve.c:
```diff
@@ -77,15 +82,73 @@ NET_BUF_POOL_DEFINE(dns_qname_pool, DNS_RESOLVER_BUF_CTR, DNS_MAX_NAME_LEN,
+{
+	if (family == AF_INET) {
+		if (net_is_ipv4_addr_mcast(&net_sin(addr)->sin_addr)) {
+			if (net_sin(addr)->sin_addr.s4_addr[3] == 252) {
```
same

tbursztyka commented on subsys/net/lib/dns/resolve.c:
```diff
@@ -77,15 +82,73 @@ NET_BUF_POOL_DEFINE(dns_qname_pool, DNS_RESOLVER_BUF_CTR, DNS_MAX_NAME_LEN,
+		if (net_is_ipv6_addr_mcast(&net_sin6(addr)->sin6_addr)) {
+			if (net_sin6(addr)->sin6_addr.s6_addr[15] == 0x03) {
+				return true;
+			} else {
```
same

tbursztyka commented on subsys/net/lib/dns/resolve.c:
```diff
@@ -787,7 +882,20 @@ int dns_resolve_name(struct dns_resolve_context *ctx,
+		 * LLMNR multicast address unless it is a mDNS query.
+		 */
+		if (!mdns_query) {
+			if (IS_ENABLED(CONFIG_LLMNR_RESOLVER)) {
```
merge that if into the first one (less imbrication then)

tbursztyka commented on subsys/net/lib/dns/llmnr_responder.c:
```diff
@@ -0,0 +1,610 @@
+	socklen_t dst_len;
+	int ret;
+
+	if (net_pkt_family(pkt) == AF_INET) {
```
looks like the code below could be factorized quite a bit actually, getting the addr according to query type mostly.

jukkar replied:
Yes, this was the tricky part of the code. Looks a bit hairy indeed. Will investigate.

tbursztyka commented on subsys/net/lib/dns/llmnr_responder.c:
```diff
@@ -0,0 +1,610 @@
+	int ret, ok = 0;
+
+#if defined(CONFIG_NET_IPV6)
+	do {
```
why the do { } while() ? {} would be sufficient. Moreover you use a goto (and not a break) inside. Either way: do/while + break or {} + goto.

tbursztyka commented on subsys/net/lib/dns/llmnr_responder.c:
```diff
@@ -0,0 +1,610 @@
+#endif /* CONFIG_NET_IPV6 */
+
+#if defined(CONFIG_NET_IPV4)
+	do {
```
same here

jukkar commented:
Updated according to comments.

pfalcon commented on samples/net/dns_resolve/README.rst:
```diff
@@ -111,6 +115,27 @@ For testing mDNS, use Avahi script in net-tools project:
+A Zephyr host needs a hostname assigned to it so that it can respond to a DNS
+query. Note that the hostname should not have any dots in it.
+
+Also DNS support and LLMNR responder support needs to be enabled in the
```
I'd suggest to remove this sentence. They need to be enabled, and are enabled by the config snippet above.

jukkar replied:
Sure, makes sense.

pfalcon commented on subsys/net/ip/net_shell.c:
```diff
@@ -1413,7 +1413,8 @@ static void print_dns_info(struct dns_resolve_context *ctx)
 	printk("DNS servers:\n");
 
-	for (i = 0; i < CONFIG_DNS_RESOLVER_MAX_SERVERS + MDNS_SERVER_COUNT; i++) {
+	for (i = 0; i < CONFIG_DNS_RESOLVER_MAX_SERVERS +
```
@tbursztyka, @jukkar: Re: this comment: https://github.com/zephyrproject-rtos/zephyr/pull/7251/files#r198456907

That's where it gets a bit hairy, yeah.

jukkar replied:
Hmm, I missed this change. Will fix.

pfalcon reviewed (approved):
LGTM. There're some cosmetic issues, I'll let @tbursztyka decide how important they're to resolve.

jukkar commented on subsys/net/lib/dns/Kconfig:
```diff
@@ -20,6 +20,17 @@ config MDNS_RESOLVER
+	  support. See RFC 4795 for details. LLMNR is typically used by Windows
+	  hosts. If you enable this option, then the DNS requests are ONLY sent
+	  to LLMNR well known multicast address 224.0.0.252:5355 or
+	  [ff02::1:3]:5355 and other DNS server addresses are ignored.
```
It gets quite complicated if we would need to maintain two DNS caches. The RFC says that `In order to prevent responses to LLMNR queries from polluting the DNS cache, LLMNR implementations MUST use a distinct, isolated cache for LLMNR on each interface.`

jukkar commented on subsys/net/lib/dns/Kconfig:
```diff
@@ -161,3 +172,51 @@ config MDNS_RESOLVER_ADDITIONAL_BUF_CTR
+	  It will listen well-known address ff02::1:3 and 224.0.0.252.
+	  Currently this only returns IP address information.
+	  You must set CONFIG_NET_HOSTNAME to some meaningful value and
+	  then LLMNR will start to respond to <hostname> LLMNR queries.
```
I think using <> is quite typical in examples. Also that is used in mDNS entry so I followed suit.

jukkar commented:
@pfalcon comments addressed

jukkar commented on subsys/net/lib/dns/llmnr_responder.c:
```diff
@@ -0,0 +1,650 @@
+	if (qtype == DNS_RR_TYPE_A) {
+		/* Select proper address according to destination */
+		addr = get_ipv4_src(net_pkt_iface(pkt),
+				    &net_sin(dst)->sin_addr);
```
Note that we might be getting AAAA query using IPv4, and A query using IPv6. Then depending on how the system is configured, we need to do correct thing. The original if/then/else was looking a bit too hairy, but it should be much more readable now.

tbursztyka reviewed (approved):
let's go with it
